browser/ui/views/brave_shields/cookie_list_opt_in_browsertest.cc

```diff
@@ -0,0 +1,188 @@
+  ASSERT_TRUE(web_contents);
+
+  base::RunLoop run_loop;
+  PrefChangeRegistrar pref_observer;
+  pref_observer.Init(g_browser_process->local_state());
+  pref_observer.Add(
+      prefs::kAdBlockRegionalFilters,
+      base::BindLambdaForTesting([this, &run_loop]() {
+        if (regional_service_manager()->IsFilterListEnabled(kCookieListUuid)) {
+          run_loop.Quit();
+        }
+      }));
+
+  ASSERT_EQ(true, content::EvalJs(web_contents, R"js(
+    new Promise((resolve) => {
+      setInterval(() => {
+        const elem = document.querySelector('.opt-in-action button')
+        if (elem) {
+          resolve(elem)
+        }
+      }, 30)
+    }).then((elem) => {
+      elem.click()
+      return true
+    })
+  )js"));
+
+  run_loop.Run();
+
+  EXPECT_TRUE(regional_service_manager()->IsFilterListEnabled(kCookieListUuid));
+
+  ASSERT_TRUE(
+      AddTabAtIndex(1, GURL(url::kAboutBlankURL), ui::PAGE_TRANSITION_TYPED));
+
+  WaitForLoadStop(browser()->tab_strip_model()->GetActiveWebContents());
+
+  EXPECT_FALSE(GetBubbleWebContents());
+}
+
+class CookieListOptInPreEnabledBrowserTest : public CookieListOptInBrowserTest {
+ public:
+  ~CookieListOptInPreEnabledBrowserTest() override = default;
+
+ protected:
+  void SetUpLocalState() override {
+    CookieListOptInBrowserTest::SetUpLocalState();
+    regional_service_manager()->EnableFilterList(kCookieListUuid, true);
+  }
+};
+
+IN_PROC_BROWSER_TEST_F(CookieListOptInPreEnabledBrowserTest, AlreadyEnabled) {
+  EXPECT_FALSE(GetBubbleWebContents());
+}
+
+IN_PROC_BROWSER_TEST_F(CookieListOptInBrowserTest, PRE_MultipleWindowRestore) {
+  EXPECT_TRUE(GetBubbleWebContents());
+
+  // Create another browser window.
+  CreateBrowser(browser()->profile());
+
+  // Before closing, reset the "show" pref so that the opt-in will be displayed
+  // again.
+  g_browser_process->local_state()->SetBoolean(
+      prefs::kAdBlockCookieListOptInShown, false);
+}
+
+IN_PROC_BROWSER_TEST_F(CookieListOptInBrowserTest, MultipleWindowRestore) {
+  // Count the number of windows that are showing the bubble.
+  auto* browser_list = BrowserList::GetInstance();
+  auto count = std::count_if(
+      browser_list->begin(), browser_list->end(), [](Browser* browser) {
+        if (auto* host = CookieListOptInBubbleHost::FromBrowser(browser)) {
+          if (host->GetBubbleWebContentsForTesting()) {
+            return true;
+          }
+        }
+        return false;
+      });
+
+  // Only one browser window should be showing the bubble on restore.
+  EXPECT_EQ(count, 1);
+}
+
+class CookieListOptInFirstRunBrowserTest : public CookieListOptInBrowserTest {
+ public:
+  ~CookieListOptInFirstRunBrowserTest() override = default;
+
+ protected:
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    command_line->AppendSwitch(switches::kForceFirstRun);
+  }
+};
+
+IN_PROC_BROWSER_TEST_F(CookieListOptInFirstRunBrowserTest, FirstRun) {
+  EXPECT_FALSE(GetBubbleWebContents());
```

Comment:
can we also wait for session restore, ensure the active tab has finished loading and `kAdBlockCookieListOptInShown` is false?

Reply:
Updated to wait for restore and added check for `kAdBlockCookieListOptInShown`. Browser tests automatically wait for the first tab to finish loading (see [here](https://source.chromium.org/chromium/chromium/src/+/main:chrome/test/base/in_process_browser_test.cc;l=750;drc=5ef9498cd3f6e965991dacce1333ac6a7050c26a;bpv=1;bpt=1))